// @ston-fi/api/dist/cjs/chunk-EHLYTEOW.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _chunkJ2KGXBX3cjs = require('./chunk-J2KGXBX3.cjs');


var _chunk7HHGQJQGcjs = require('./chunk-7HHGQJQG.cjs');


var _chunkFW4WJKIDcjs = require('./chunk-FW4WJKID.cjs');





var _chunkAHEA4IJVcjs = require('./chunk-AHEA4IJV.cjs');

// src/client/apiClient.ts
var _ofetch = require('ofetch');
var StonApiClient = class {
  constructor(options) {
    var _a, _b;
    const baseUrl = new URL(
      (_b = (_a = options == null ? void 0 : options.baseURL) != null ? _a : options == null ? void 0 : options.baseUrl) != null ? _b : "https://api.ston.fi"
    );
    const baseQuery = [...new URLSearchParams(baseUrl.search)].reduce(
      // biome-ignore lint/performance/noAccumulatingSpread: it's ok here
      (acc, [key, value]) => _chunkAHEA4IJVcjs.__spreadProps.call(void 0, _chunkAHEA4IJVcjs.__spreadValues.call(void 0, {}, acc), { [key]: value }),
      {}
    );
    this.apiFetch = _ofetch.ofetch.create({
      baseURL: `${baseUrl.origin}${baseUrl.pathname}`,
      query: baseQuery
    });
  }
  getAsset(assetAddress) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/assets/{assetAddress}", {
            method: "GET",
            query: { assetAddress }
          })
        )
      ).asset;
    });
  }
  getAssets() {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/assets", {
            method: "GET"
          })
        )
      ).assetList;
    });
  }
  queryAssets(_a) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      var _b = _a, {
        unconditionalAssets: unconditionalAsset
      } = _b, query = _chunkAHEA4IJVcjs.__objRest.call(void 0, _b, [
        "unconditionalAssets"
      ]);
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/assets/query", {
            method: "POST",
            query: _chunkAHEA4IJVcjs.__spreadProps.call(void 0, _chunkAHEA4IJVcjs.__spreadValues.call(void 0, {}, query), {
              unconditionalAsset
            })
          })
        )
      ).assetList;
    });
  }
  searchAssets(_c) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      var _d = _c, {
        unconditionalAssets: unconditionalAsset
      } = _d, query = _chunkAHEA4IJVcjs.__objRest.call(void 0, _d, [
        "unconditionalAssets"
      ]);
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/assets/search", {
            method: "POST",
            query: _chunkAHEA4IJVcjs.__spreadProps.call(void 0, _chunkAHEA4IJVcjs.__spreadValues.call(void 0, {}, query), {
              unconditionalAsset
            })
          })
        )
      ).assetList;
    });
  }
  getFarm(farmAddress) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/farms/{farmAddress}", {
            method: "GET",
            query: { farmAddress }
          })
        )
      ).farm;
    });
  }
  getFarms(query) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/farms", {
            method: "GET",
            query
          })
        )
      ).farmList;
    });
  }
  getFarmsByPool(poolAddress) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/farms_by_pool/{poolAddress}", {
            method: "GET",
            query: { poolAddress }
          })
        )
      ).farmList;
    });
  }
  getSwapPairs(query) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/markets", {
            method: "GET",
            query
          })
        )
      ).pairs;
    });
  }
  getSwapStatus(query) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/swap/status", {
            method: "GET",
            query
          })
        )
      );
    });
  }
  getPool(data) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/pools/{poolAddress}", {
            method: "GET",
            query: typeof data === "string" ? { poolAddress: data } : data
          })
        )
      ).pool;
    });
  }
  getPools(query) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/pools", {
            method: "GET",
            query
          })
        )
      ).poolList;
    });
  }
  getPoolsByAssetPair(query) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, 
            "/v1/pools/by_market/{asset0Address}/{asset1Address}",
            {
              method: "GET",
              query
            }
          )
        )
      ).poolList;
    });
  }
  queryPools(_e) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      var _f = _e, {
        unconditionalAssets: unconditionalAsset
      } = _f, query = _chunkAHEA4IJVcjs.__objRest.call(void 0, _f, [
        "unconditionalAssets"
      ]);
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/pool/query", {
            method: "POST",
            query: _chunkAHEA4IJVcjs.__spreadProps.call(void 0, _chunkAHEA4IJVcjs.__spreadValues.call(void 0, {}, query), {
              unconditionalAsset
            })
          })
        )
      ).poolList;
    });
  }
  simulateSwap(_g) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      var _h = _g, {
        offerUnits: units
      } = _h, query = _chunkAHEA4IJVcjs.__objRest.call(void 0, _h, [
        "offerUnits"
      ]);
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/swap/simulate", {
            method: "POST",
            query: _chunkAHEA4IJVcjs.__spreadProps.call(void 0, _chunkAHEA4IJVcjs.__spreadValues.call(void 0, {}, query), {
              units
            })
          })
        )
      );
    });
  }
  simulateReverseSwap(_i) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      var _j = _i, {
        askUnits: units
      } = _j, query = _chunkAHEA4IJVcjs.__objRest.call(void 0, _j, [
        "askUnits"
      ]);
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/reverse_swap/simulate", {
            method: "POST",
            query: _chunkAHEA4IJVcjs.__spreadProps.call(void 0, _chunkAHEA4IJVcjs.__spreadValues.call(void 0, {}, query), {
              units
            })
          })
        )
      );
    });
  }
  simulateLiquidityProvision(query) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/liquidity_provision/simulate", {
            method: "POST",
            query
          })
        )
      );
    });
  }
  getJettonWalletAddress(query) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/jetton/{jettonAddress}/address", {
            method: "GET",
            query
          })
        )
      ).address;
    });
  }
  getWalletAsset(query) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, 
            "/v1/wallets/{walletAddress}/assets/{assetAddress}",
            {
              method: "GET",
              query
            }
          )
        )
      ).asset;
    });
  }
  getWalletAssets(walletAddress) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/wallets/{walletAddress}/assets", {
            method: "GET",
            query: { walletAddress }
          })
        )
      ).assetList;
    });
  }
  getWalletFarm(query) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/wallets/{walletAddress}/farms/{farmAddress}", {
            method: "GET",
            query
          })
        )
      ).farm;
    });
  }
  getWalletFarms(data) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/wallets/{walletAddress}/farms", {
            method: "GET",
            query: typeof data === "string" ? { walletAddress: data } : data
          })
        )
      ).farmList;
    });
  }
  getWalletPool(query) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/wallets/{walletAddress}/pools/{poolAddress}", {
            method: "GET",
            query
          })
        )
      ).pool;
    });
  }
  getWalletPools(data) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/wallets/{walletAddress}/pools", {
            method: "GET",
            query: typeof data === "string" ? { walletAddress: data } : data
          })
        )
      ).poolList;
    });
  }
  getWalletOperations(_k) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      var _l = _k, {
        since,
        until
      } = _l, query = _chunkAHEA4IJVcjs.__objRest.call(void 0, _l, [
        "since",
        "until"
      ]);
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/wallets/{walletAddress}/operations", {
            method: "GET",
            query: _chunkAHEA4IJVcjs.__spreadProps.call(void 0, _chunkAHEA4IJVcjs.__spreadValues.call(void 0, {}, query), {
              since: _chunk7HHGQJQGcjs.normalizeDate.call(void 0, since),
              until: _chunk7HHGQJQGcjs.normalizeDate.call(void 0, until)
            })
          })
        )
      ).operations;
    });
  }
  getOperations(_0) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, arguments, function* ({ since, until }) {
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/stats/operations", {
            method: "GET",
            query: {
              since: _chunk7HHGQJQGcjs.normalizeDate.call(void 0, since),
              until: _chunk7HHGQJQGcjs.normalizeDate.call(void 0, until)
            }
          })
        )
      ).operations;
    });
  }
  getRouters(query) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/routers", {
            method: "GET",
            query
          })
        )
      ).routerList;
    });
  }
  getRouter(routerAddress) {
    return _chunkAHEA4IJVcjs.__async.call(void 0, this, null, function* () {
      return _chunkJ2KGXBX3cjs.normalizeResponse.call(void 0, 
        yield this.apiFetch(
          ..._chunkFW4WJKIDcjs.normalizeRequest.call(void 0, "/v1/routers/{routerAddress}", {
            method: "GET",
            query: {
              routerAddress
            }
          })
        )
      ).router;
    });
  }
};



exports.StonApiClient = StonApiClient;
