// @ston-fi/api/dist/cjs/chunk-5ZR4QMSK.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});// src/client/types/asset.ts
var AssetKind = {
  Ton: "Ton",
  Wton: "Wton",
  Jetton: "Jetton"
};
var AssetTag = {
  Essential: "asset:essential",
  Blacklisted: "asset:blacklisted",
  Deprecated: "asset:deprecated",
  DefaultSymbol: "asset:default_symbol",
  LiquidityNo: "asset:liquidity:no",
  LiquidityMedium: "asset:liquidity:medium",
  LiquidityLow: "asset:liquidity:low",
  LiquidityHigh: "asset:liquidity:high",
  LiquidityVeryHigh: "asset:liquidity:very_high",
  Popular: "asset:popular",
  WalletHasBalance: "asset:wallet_has_balance",
  WalletHasLiquidityInPool: "asset:wallet_has_liquidity_in_pool",
  Taxable: "asset:taxable",
  /** @deprecated use LiquidityNo instead */
  NoLiquidity: "no_liquidity",
  /** @deprecated use LiquidityLow instead */
  LowLiquidity: "low_liquidity",
  /** @deprecated use LiquidityHigh instead */
  HighLiquidity: "high_liquidity"
};




exports.AssetKind = AssetKind; exports.AssetTag = AssetTag;
