// @ston-fi/sdk/dist/chunk-GLLFDRSY.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); var _class;

var _chunkBEV6I4FTcjs = require('./chunk-BEV6I4FT.cjs');


var _chunkLYVQCGUPcjs = require('./chunk-LYVQCGUP.cjs');


var _chunkYFCQJXY3cjs = require('./chunk-YFCQJXY3.cjs');


var _chunkWF6EZ2CWcjs = require('./chunk-WF6EZ2CW.cjs');


var _chunkZJQ75LD3cjs = require('./chunk-ZJQ75LD3.cjs');

// src/contracts/dex/v2_2/router/WStableRouterV2_2.ts
var WStableRouterV2_2 = (_class = class extends _chunkYFCQJXY3cjs.WStableRouterV2_1 {
  static __initStatic() {this.version = _chunkZJQ75LD3cjs.DEX_VERSION.v2_2}
  async getPool(provider, params) {
    const poolAddress = await this.getPoolAddressByJettonMinters(
      provider,
      params
    );
    return _chunkBEV6I4FTcjs.WStablePoolV2_2.create(poolAddress);
  }
  async getVault(provider, params) {
    const tokenMinter = provider.open(_chunkWF6EZ2CWcjs.JettonMinter.create(params.tokenMinter));
    const vaultAddress = await this.getVaultAddress(provider, {
      user: params.user,
      tokenWallet: await tokenMinter.getWalletAddress(this.address)
    });
    return _chunkLYVQCGUPcjs.VaultV2_2.create(vaultAddress);
  }
}, _class.__initStatic(), _class);



exports.WStableRouterV2_2 = WStableRouterV2_2;
