// @ston-fi/sdk/dist/chunk-A2XXZLQM.js
import {
  StablePoolV2_1
} from "./chunk-OXC3GXIH.js";
import {
  BaseRouterV2_1
} from "./chunk-B5WQRUWB.js";
import {
  DEX_TYPE
} from "./chunk-4KVHWM75.js";

// src/contracts/dex/v2_1/router/StableRouterV2_1.ts
import { toNano } from "@ton/ton";
var StableRouterV2_1 = class _StableRouterV2_1 extends BaseRouterV2_1 {
  static dexType = DEX_TYPE.Stable;
  static gasConstants = {
    ...BaseRouterV2_1.gasConstants,
    swapJettonToJetton: {
      gasAmount: toNano("0.329"),
      forwardGasAmount: toNano("0.269")
    },
    swapJettonToTon: {
      gasAmount: toNano("0.329"),
      forwardGasAmount: toNano("0.269")
    },
    swapTonToJetton: {
      forwardGasAmount: toNano("0.329")
    }
  };
  constructor(address, { gasConstants, ...options } = {}) {
    super(address, {
      ...options,
      gasConstants: {
        ..._StableRouterV2_1.gasConstants,
        ...gasConstants
      }
    });
  }
  async getPool(provider, params) {
    const poolAddress = await this.getPoolAddressByJettonMinters(
      provider,
      params
    );
    return StablePoolV2_1.create(poolAddress);
  }
};

export {
  StableRouterV2_1
};
