// @ston-fi/sdk/dist/chunk-YFCQJXY3.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); var _class;

var _chunkGOIXG6S3cjs = require('./chunk-GOIXG6S3.cjs');


var _chunkI5N4ORCVcjs = require('./chunk-I5N4ORCV.cjs');


var _chunkZJQ75LD3cjs = require('./chunk-ZJQ75LD3.cjs');

// src/contracts/dex/v2_1/router/WStableRouterV2_1.ts
var _ton = require('@ton/ton');
var WStableRouterV2_1 = (_class = class _WStableRouterV2_1 extends _chunkI5N4ORCVcjs.BaseRouterV2_1 {
  static __initStatic() {this.dexType = _chunkZJQ75LD3cjs.DEX_TYPE.WStable}
  static __initStatic2() {this.gasConstants = {
    ..._chunkI5N4ORCVcjs.BaseRouterV2_1.gasConstants,
    swapJettonToJetton: {
      gasAmount: _ton.toNano.call(void 0, "0.479"),
      forwardGasAmount: _ton.toNano.call(void 0, "0.419")
    },
    swapJettonToTon: {
      gasAmount: _ton.toNano.call(void 0, "0.479"),
      forwardGasAmount: _ton.toNano.call(void 0, "0.419")
    },
    swapTonToJetton: {
      forwardGasAmount: _ton.toNano.call(void 0, "0.479")
    }
  }}
  constructor(address, { gasConstants, ...options } = {}) {
    super(address, {
      ...options,
      gasConstants: {
        ..._WStableRouterV2_1.gasConstants,
        ...gasConstants
      }
    });
  }
  async getPool(provider, params) {
    const poolAddress = await this.getPoolAddressByJettonMinters(
      provider,
      params
    );
    return _chunkGOIXG6S3cjs.WStablePoolV2_1.create(poolAddress);
  }
}, _class.__initStatic(), _class.__initStatic2(), _class);



exports.WStableRouterV2_1 = WStableRouterV2_1;
