// @ston-fi/sdk/dist/chunk-AIIUO6JT.js
import {
  LpAccountV2_1
} from "./chunk-3KKQOYNT.js";
import {
  DEX_OP_CODES
} from "./chunk-CWGXHOMU.js";
import {
  JettonMinter
} from "./chunk-S2MQPWHB.js";
import {
  JettonWallet
} from "./chunk-6OHMCTKH.js";
import {
  DEX_VERSION
} from "./chunk-4KVHWM75.js";
import {
  toAddress
} from "./chunk-HNMPFVZW.js";

// src/contracts/dex/v2_1/pool/BasePoolV2_1.ts
import {
  beginCell,
  toNano
} from "@ton/ton";
var BasePoolV2_1 = class _BasePoolV2_1 extends JettonMinter {
  static version = DEX_VERSION.v2_1;
  static gasConstants = {
    collectFees: toNano("0.4"),
    burn: toNano("0.8")
  };
  gasConstants;
  constructor(address, { gasConstants, ...options } = {}) {
    super(address, options);
    this.gasConstants = {
      ..._BasePoolV2_1.gasConstants,
      ...gasConstants
    };
  }
  async createCollectFeesBody(params) {
    return beginCell().storeUint(DEX_OP_CODES.COLLECT_FEES, 32).storeUint(params?.queryId ?? 0, 64).endCell();
  }
  async getCollectFeeTxParams(provider, params) {
    const to = this.address;
    const body = await this.createCollectFeesBody({
      queryId: params?.queryId
    });
    const value = BigInt(params?.gasAmount ?? this.gasConstants.collectFees);
    return { to, value, body };
  }
  async sendCollectFees(provider, via, params) {
    const txParams = await this.getCollectFeeTxParams(provider, params);
    return via.send(txParams);
  }
  async createBurnBody(params) {
    return beginCell().storeUint(DEX_OP_CODES.BURN, 32).storeUint(params?.queryId ?? 0, 64).storeCoins(BigInt(params.amount)).storeAddress(null).storeMaybeRef(params.dexCustomPayload).endCell();
  }
  async getBurnTxParams(provider, params) {
    const [to, body] = await Promise.all([
      this.getWalletAddress(provider, params.userWalletAddress),
      this.createBurnBody({
        amount: params.amount,
        dexCustomPayload: params.dexCustomPayload,
        queryId: params.queryId
      })
    ]);
    const value = BigInt(params.gasAmount ?? this.gasConstants.burn);
    return { to, value, body };
  }
  async sendBurn(provider, via, params) {
    const txParams = await this.getBurnTxParams(provider, params);
    return via.send(txParams);
  }
  async getPoolType(provider) {
    const result = await provider.get("get_pool_type", []);
    return result.stack.readString();
  }
  async getLpAccountAddress(provider, params) {
    const result = await provider.get("get_lp_account_address", [
      {
        type: "slice",
        cell: beginCell().storeAddress(toAddress(params.ownerAddress)).endCell()
      }
    ]);
    return result.stack.readAddress();
  }
  async getLpAccount(provider, params) {
    const lpAccountAddress = await this.getLpAccountAddress(provider, params);
    return LpAccountV2_1.create(lpAccountAddress);
  }
  async getJettonWallet(provider, params) {
    const jettonWalletAddress = await this.getWalletAddress(
      provider,
      params.ownerAddress
    );
    return JettonWallet.create(jettonWalletAddress);
  }
  async getPoolData(provider) {
    const data = await this.implGetPoolData(provider);
    return data.commonPoolData;
  }
  async implGetPoolData(provider) {
    const result = await provider.get("get_pool_data", []);
    return {
      commonPoolData: {
        isLocked: result.stack.readBoolean(),
        routerAddress: result.stack.readAddress(),
        totalSupplyLP: result.stack.readBigNumber(),
        reserve0: result.stack.readBigNumber(),
        reserve1: result.stack.readBigNumber(),
        token0WalletAddress: result.stack.readAddress(),
        token1WalletAddress: result.stack.readAddress(),
        lpFee: result.stack.readBigNumber(),
        protocolFee: result.stack.readBigNumber(),
        protocolFeeAddress: result.stack.readAddressOpt(),
        collectedToken0ProtocolFee: result.stack.readBigNumber(),
        collectedToken1ProtocolFee: result.stack.readBigNumber()
      },
      stack: result.stack
    };
  }
};

export {
  BasePoolV2_1
};
