// @ston-fi/sdk/dist/chunk-MLD4WHXP.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); var _class; var _class2;

var _chunkWNOITHHMcjs = require('./chunk-WNOITHHM.cjs');


var _chunkVBK5IA22cjs = require('./chunk-VBK5IA22.cjs');


var _chunkYFCQJXY3cjs = require('./chunk-YFCQJXY3.cjs');


var _chunkM25E5KQRcjs = require('./chunk-M25E5KQR.cjs');


var _chunkWQUCOAKWcjs = require('./chunk-WQUCOAKW.cjs');


var _chunk7ENVDM26cjs = require('./chunk-7ENVDM26.cjs');


var _chunkGIDOXPICcjs = require('./chunk-GIDOXPIC.cjs');


var _chunkGOIXG6S3cjs = require('./chunk-GOIXG6S3.cjs');


var _chunkI5N4ORCVcjs = require('./chunk-I5N4ORCV.cjs');


var _chunkRPU6RIXMcjs = require('./chunk-RPU6RIXM.cjs');


var _chunkF6QEZUEHcjs = require('./chunk-F6QEZUEH.cjs');


var _chunkARDWDSLZcjs = require('./chunk-ARDWDSLZ.cjs');


var _chunkRRI66KHLcjs = require('./chunk-RRI66KHL.cjs');

// src/contracts/dex/v2_1/index.ts
var RouterV2_1 = (_class = class extends _chunkI5N4ORCVcjs.BaseRouterV2_1 {
  static __initStatic() {this.CPI = _chunkM25E5KQRcjs.CPIRouterV2_1}
  static __initStatic2() {this.Stable = _chunkWNOITHHMcjs.StableRouterV2_1}
  static __initStatic3() {this.WCPI = _chunkVBK5IA22cjs.WCPIRouterV2_1}
  static __initStatic4() {this.WStable = _chunkYFCQJXY3cjs.WStableRouterV2_1}
}, _class.__initStatic(), _class.__initStatic2(), _class.__initStatic3(), _class.__initStatic4(), _class);
var PoolV2_1 = (_class2 = class extends _chunkF6QEZUEHcjs.BasePoolV2_1 {
  static __initStatic5() {this.CPI = _chunkWQUCOAKWcjs.CPIPoolV2_1}
  static __initStatic6() {this.Stable = _chunk7ENVDM26cjs.StablePoolV2_1}
  static __initStatic7() {this.WCPI = _chunkGIDOXPICcjs.WCPIPoolV2_1}
  static __initStatic8() {this.WStable = _chunkGOIXG6S3cjs.WStablePoolV2_1}
}, _class2.__initStatic5(), _class2.__initStatic6(), _class2.__initStatic7(), _class2.__initStatic8(), _class2);
var DEX = {
  Router: RouterV2_1,
  Pool: PoolV2_1,
  LpAccount: _chunkARDWDSLZcjs.LpAccountV2_1,
  Vault: _chunkRPU6RIXMcjs.VaultV2_1,
  pTON: _chunkRRI66KHLcjs.PtonV2_1
};





exports.RouterV2_1 = RouterV2_1; exports.PoolV2_1 = PoolV2_1; exports.DEX = DEX;
