// @ston-fi/sdk/dist/chunk-C4LRQUDR.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); var _class; var _class2;

var _chunkJATP75MMcjs = require('./chunk-JATP75MM.cjs');


var _chunkLBMSMRICcjs = require('./chunk-LBMSMRIC.cjs');


var _chunkGLLFDRSYcjs = require('./chunk-GLLFDRSY.cjs');


var _chunkXUOWF5LZcjs = require('./chunk-XUOWF5LZ.cjs');


var _chunkFSVUI5QVcjs = require('./chunk-FSVUI5QV.cjs');


var _chunkBEV6I4FTcjs = require('./chunk-BEV6I4FT.cjs');


var _chunkFD62DY52cjs = require('./chunk-FD62DY52.cjs');


var _chunkIDIAIBUUcjs = require('./chunk-IDIAIBUU.cjs');


var _chunkIOSGKCC2cjs = require('./chunk-IOSGKCC2.cjs');


var _chunkYN73SU6Rcjs = require('./chunk-YN73SU6R.cjs');


var _chunkLYVQCGUPcjs = require('./chunk-LYVQCGUP.cjs');


var _chunkUJRPXSQBcjs = require('./chunk-UJRPXSQB.cjs');


var _chunkRRI66KHLcjs = require('./chunk-RRI66KHL.cjs');

// src/contracts/dex/v2_2/index.ts
var RouterV2_2 = (_class = class extends _chunkFD62DY52cjs.BaseRouterV2_2 {
  static __initStatic() {this.CPI = _chunkIOSGKCC2cjs.CPIRouterV2_2}
  static __initStatic2() {this.Stable = _chunkJATP75MMcjs.StableRouterV2_2}
  static __initStatic3() {this.WCPI = _chunkLBMSMRICcjs.WCPIRouterV2_2}
  static __initStatic4() {this.WStable = _chunkGLLFDRSYcjs.WStableRouterV2_2}
}, _class.__initStatic(), _class.__initStatic2(), _class.__initStatic3(), _class.__initStatic4(), _class);
var PoolV2_2 = (_class2 = class extends _chunkIDIAIBUUcjs.BasePoolV2_2 {
  static __initStatic5() {this.CPI = _chunkYN73SU6Rcjs.CPIPoolV2_2}
  static __initStatic6() {this.Stable = _chunkXUOWF5LZcjs.StablePoolV2_2}
  static __initStatic7() {this.WCPI = _chunkFSVUI5QVcjs.WCPIPoolV2_2}
  static __initStatic8() {this.WStable = _chunkBEV6I4FTcjs.WStablePoolV2_2}
}, _class2.__initStatic5(), _class2.__initStatic6(), _class2.__initStatic7(), _class2.__initStatic8(), _class2);
var DEX = {
  Router: RouterV2_2,
  Pool: PoolV2_2,
  LpAccount: _chunkUJRPXSQBcjs.LpAccountV2_2,
  Vault: _chunkLYVQCGUPcjs.VaultV2_2,
  pTON: _chunkRRI66KHLcjs.PtonV2_1
};





exports.RouterV2_2 = RouterV2_2; exports.PoolV2_2 = PoolV2_2; exports.DEX = DEX;
