// @ston-fi/sdk/dist/chunk-VBK5IA22.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); var _class;

var _chunkGIDOXPICcjs = require('./chunk-GIDOXPIC.cjs');


var _chunkI5N4ORCVcjs = require('./chunk-I5N4ORCV.cjs');


var _chunkZJQ75LD3cjs = require('./chunk-ZJQ75LD3.cjs');

// src/contracts/dex/v2_1/router/WCPIRouterV2_1.ts
var _ton = require('@ton/ton');
var WCPIRouterV2_1 = (_class = class _WCPIRouterV2_1 extends _chunkI5N4ORCVcjs.BaseRouterV2_1 {
  static __initStatic() {this.dexType = _chunkZJQ75LD3cjs.DEX_TYPE.WCPI}
  static __initStatic2() {this.gasConstants = {
    ..._chunkI5N4ORCVcjs.BaseRouterV2_1.gasConstants,
    swapJettonToJetton: {
      gasAmount: _ton.toNano.call(void 0, "0.319"),
      forwardGasAmount: _ton.toNano.call(void 0, "0.259")
    },
    swapJettonToTon: {
      gasAmount: _ton.toNano.call(void 0, "0.319"),
      forwardGasAmount: _ton.toNano.call(void 0, "0.259")
    },
    swapTonToJetton: {
      forwardGasAmount: _ton.toNano.call(void 0, "0.319")
    }
  }}
  constructor(address, { gasConstants, ...options } = {}) {
    super(address, {
      ...options,
      gasConstants: {
        ..._WCPIRouterV2_1.gasConstants,
        ...gasConstants
      }
    });
  }
  async getPool(provider, params) {
    const poolAddress = await this.getPoolAddressByJettonMinters(
      provider,
      params
    );
    return _chunkGIDOXPICcjs.WCPIPoolV2_1.create(poolAddress);
  }
}, _class.__initStatic(), _class.__initStatic2(), _class);



exports.WCPIRouterV2_1 = WCPIRouterV2_1;
