// @ston-fi/sdk/dist/chunk-UGZE6JY6.js
import {
  WCPIPoolV2_1
} from "./chunk-GMIWAM73.js";
import {
  BaseRouterV2_1
} from "./chunk-B5WQRUWB.js";
import {
  DEX_TYPE
} from "./chunk-4KVHWM75.js";

// src/contracts/dex/v2_1/router/WCPIRouterV2_1.ts
import { toNano } from "@ton/ton";
var WCPIRouterV2_1 = class _WCPIRouterV2_1 extends BaseRouterV2_1 {
  static dexType = DEX_TYPE.WCPI;
  static gasConstants = {
    ...BaseRouterV2_1.gasConstants,
    swapJettonToJetton: {
      gasAmount: toNano("0.319"),
      forwardGasAmount: toNano("0.259")
    },
    swapJettonToTon: {
      gasAmount: toNano("0.319"),
      forwardGasAmount: toNano("0.259")
    },
    swapTonToJetton: {
      forwardGasAmount: toNano("0.319")
    }
  };
  constructor(address, { gasConstants, ...options } = {}) {
    super(address, {
      ...options,
      gasConstants: {
        ..._WCPIRouterV2_1.gasConstants,
        ...gasConstants
      }
    });
  }
  async getPool(provider, params) {
    const poolAddress = await this.getPoolAddressByJettonMinters(
      provider,
      params
    );
    return WCPIPoolV2_1.create(poolAddress);
  }
};

export {
  WCPIRouterV2_1
};
