// @ston-fi/sdk/dist/chunk-MTUDFKLM.js
import {
  WStablePoolV2_1
} from "./chunk-C7SF5WJK.js";
import {
  BaseRouterV2_1
} from "./chunk-B5WQRUWB.js";
import {
  DEX_TYPE
} from "./chunk-4KVHWM75.js";

// src/contracts/dex/v2_1/router/WStableRouterV2_1.ts
import { toNano } from "@ton/ton";
var WStableRouterV2_1 = class _WStableRouterV2_1 extends BaseRouterV2_1 {
  static dexType = DEX_TYPE.WStable;
  static gasConstants = {
    ...BaseRouterV2_1.gasConstants,
    swapJettonToJetton: {
      gasAmount: toNano("0.479"),
      forwardGasAmount: toNano("0.419")
    },
    swapJettonToTon: {
      gasAmount: toNano("0.479"),
      forwardGasAmount: toNano("0.419")
    },
    swapTonToJetton: {
      forwardGasAmount: toNano("0.479")
    }
  };
  constructor(address, { gasConstants, ...options } = {}) {
    super(address, {
      ...options,
      gasConstants: {
        ..._WStableRouterV2_1.gasConstants,
        ...gasConstants
      }
    });
  }
  async getPool(provider, params) {
    const poolAddress = await this.getPoolAddressByJettonMinters(
      provider,
      params
    );
    return WStablePoolV2_1.create(poolAddress);
  }
};

export {
  WStableRouterV2_1
};
