// @ston-fi/sdk/dist/chunk-A6YGVTP3.js
import {
  StableRouterV2_2
} from "./chunk-Q24KESNX.js";
import {
  WCPIRouterV2_2
} from "./chunk-KXZ5O3JL.js";
import {
  WStableRouterV2_2
} from "./chunk-HKXZUGHL.js";
import {
  StablePoolV2_2
} from "./chunk-OE3ALABD.js";
import {
  WCPIPoolV2_2
} from "./chunk-ATD4XEMR.js";
import {
  WStablePoolV2_2
} from "./chunk-ZSYXOVW7.js";
import {
  BaseRouterV2_2
} from "./chunk-ZTKPOIAK.js";
import {
  BasePoolV2_2
} from "./chunk-KLN4CPQ5.js";
import {
  CPIRouterV2_2
} from "./chunk-TZ62L7NV.js";
import {
  CPIPoolV2_2
} from "./chunk-SNCPMR4R.js";
import {
  VaultV2_2
} from "./chunk-YOE5IHCZ.js";
import {
  LpAccountV2_2
} from "./chunk-YQZMKMIC.js";
import {
  PtonV2_1
} from "./chunk-5C3RWSYW.js";

// src/contracts/dex/v2_2/index.ts
var RouterV2_2 = class extends BaseRouterV2_2 {
  static CPI = CPIRouterV2_2;
  static Stable = StableRouterV2_2;
  static WCPI = WCPIRouterV2_2;
  static WStable = WStableRouterV2_2;
};
var PoolV2_2 = class extends BasePoolV2_2 {
  static CPI = CPIPoolV2_2;
  static Stable = StablePoolV2_2;
  static WCPI = WCPIPoolV2_2;
  static WStable = WStablePoolV2_2;
};
var DEX = {
  Router: RouterV2_2,
  Pool: PoolV2_2,
  LpAccount: LpAccountV2_2,
  Vault: VaultV2_2,
  pTON: PtonV2_1
};

export {
  RouterV2_2,
  PoolV2_2,
  DEX
};
